code-studio/src/js/callouts.js

```diff
@@ -112,16 +124,49 @@ module.exports = function createCallouts(callouts) {
     }
 
     if (callout.on) {
-      $(window).on(callout.on, function () {
-        if (!callout.seen && $(selector).length > 0) {
+      $(window).on(callout.on, function (e, action) {
+        if (!config.codeStudio.selector) {
+          config.codeStudio.selector = selector;
+        }
+        var lastSelector = config.codeStudio.selector;
+        $(window).trigger('prepareforcallout', [config.codeStudio]);
+        if (lastSelector !== config.codeStudio.selector && $(lastSelector).length > 0) {
+          $(lastSelector).qtip(config).qtip('destroy');
+        }
+        if ($(config.codeStudio.selector).length > 0) {
+          if (action === 'hashchange' || action === 'hashinit' || !callout.seen) {
+            $(config.codeStudio.selector).qtip(config).qtip('show');
+          }
           callout.seen = true;
-          $(selector).qtip(config).qtip('show');
         }
       });
-    } else {
+    } else if (!callout.seen) {
       $(selector).qtip(config).qtip('show');
     }
   });
+
+  // Insert a hashchange handler to detect triggercallout= hashes and fire
+  // appropriate events to open the callout
+  function detectTriggerCalloutOnHash(event) {
```

Comment:
Ooh, this is an interesting feature, and I can see the reasoning behind both approaches.

In the current implementation we're treating the hash as an action, executed on page-load or on hash-change.  This is fairly consistent with the default behavior of hashes for jumping to anchors - it's a one time "scroll-to" action, and you're always free to scroll away after the action occurs.  In that sense, having a hash show a callout seems reasonable.

This is different from the way many webapps use hashes, as a sort of bookmarkable page-state: Imagine keeping an updated `#code` or `#design` in the URL to track which tab we're on in applab.  I don't recommend we ever take that approach, since we handle state differently, and that sort of thing tends to require more centralized routing than we are set up for.  I think react-router might be doing this under the hood though; maybe worth double-checking that we don't stomp on that somehow.

Brent's makes a good point about not depending on the `window.location`.  You could do postprocessing on instructions and even create a custom syntax for this sort of callout action that would read really nice in markdown:

```
You need to [draw sprites](callout:drawSpritesCallout) and
```

But there are downsides too:

* If you want to do a callout-showing link outside of instructions, you've got to do the postprocessing on whatever element contains the new link - and if you forget to do so, those links are broken!  With hashes the functionality is available to any hyperlink, anywhere.
* You lose the action-on-page-load feature.  You could always get it back by adding a queryparam for showing a callout on load, but then you've put more logic in another place.
* And of course, changing this means Chris has to tear out some stuff and write something else.  Not to suffer the sunk-cost fallacy, but that is a small factor.

Overall, I feel like sticking with the hash-action approach is fine for now, and could be revisited if we think it's conflicting with something.  I _would_ like to see the location and history manipulation extracted into some sort of `HashAction` module that callouts.js consumes, per SRP.